[step2] 자동차 경주 과제

RoadtoS7 opened the pull request:
랜덤값에 따라서 결과가 달라지는 구조여서 중간부터 테스트 코드를 짤 수가 없더라구요ㅠㅠ
나중에 조금씩 리팩토링 하고나서도 랜덤으로 인해 앞서서 짜둔 테스트도 다 통과 실패하구요..ㅎㅎㅠ
그래도 나름 짜보려고 노력했습니다..!

처음으로 InputView와 OutputView를 적용해봤는데 잘 했는지 궁금합니다!
그리고 랜덤 넘버로 입력받아서 하는 함수는 어떻게 테스트해야하는지 궁금합니다.

고민을 조금 많이 했던 부분이 Car 클래스를 만들고 나서도 Cars클래스를 또 만든 것 인데요..!
일급 컬렉션을 적용하려고 하다 보니 Car리스트를 가지는 Cars를 따로 만드는 게 맞다는 생각이 들었습니다.
이렇게 나누어도 괜찮은지 궁금합니다!

또한 문자열 계산기와 달리 자동차 경주에서는 사용자의 입력이 잘못되면 그냥 에러가 나도록 만들었는데.. 따로 체크하는 클래스를 또 만드는 게 나을까요?

마지막으로 RaceRule 객체에서 랜덤 수를 생성하고 이 수가 4 이상이면 true를 반환하는 함수를 만들었습니다.
그리고 Car클래스 안의 run 함수 내에서 이 함수를 호출함으로써 따로 Car클래스의 run 함수를 사용하기에 앞서서 검증 로직을 써야함을 알지 못해도 램덤수가 4이상일 때만 run 함수의 로직이 실행되도록 했습니다.
랜덤수 체크 로직을 run 함수 안에 넣을까 고민했는데 이렇게 되면 Car클래스와 run 함수가 하는 일이 많아지는 것(? 같아서
나누었습니다. 이 부분도 그냥 합치는 게 나을지 궁금합니다!

코드리뷰 부탁드립니다~~

RoadtoS7 committed:
feat(Car.kt): add Car.kt

자동차 하나에 대한 클래스를 만들었다

RoadtoS7 committed:
feat(Cars.kt): add Cars.kt

자동차 여러대를 다루는 클래스를 만들었다

RoadtoS7 committed:
feat(RaceRule.kt): add RaceRule class for check if car run

차가 달릴 수 있는 규칙에 따라서 랜덤넘버를 생성하고 체크하는 함수를 가지는 클래스를 만들었다

RoadtoS7 committed:
feat(Race.kt): add Race class

차를 달리기 시키는 클래스인 Race 클래스를 만들었다

RoadtoS7 committed:
feat(InputView.kt): add InputView

사용자로부터 입력을 받기 위해 사용하는 함수를 담은 InputView를 만들었다

RoadtoS7 committed:
feat(InputView.kt): add OutputView.kt

실행결과를 출력하기 위한 함수를 담은 OutputView를 만들었다

RoadtoS7 committed:
feat(RacingCarApplication.kt): add RacingCarApplication.kt

main함수를 담은 RacingCarApplication.kt를 만들었다

RoadtoS7 committed:
test(CarTest.kt): add CarTest for testing Car class

Car클래스의 함수들을 테스트하기 위한 CarTest 테스트 클래스를 만들었다

RoadtoS7 committed:
test(CarsTest.kt): add CarsTest for testing Cars class

Cars클래스의 함수들을 테스트하기 위한 CarsTest 테스트 클래스를 만들었다

RoadtoS7 committed:
test(RaceRuleTest.kt): add RaceRuleTest for testing RaceRuleTest object

RaceRule 객체를 테스트하기 위한 RaceRuleTest 를 만들었다

RoadtoS7 committed:
test(RaceTest.kt): add RaceTest for testing RaceTest object

Race 클래스를 테스트하기 위한 RaceTest 를 만들었다

RoadtoS7 committed:
refactor(Car.kt): fix to apply ktlint

RoadtoS7 committed:
refactor(Cars.kt): fix to apply ktlint

RoadtoS7 committed:
refactor(RaceResult.kt): fix to apply ktlint

malibinYun commented on src/main/kotlin/racingcar/Cars.kt:
```diff
@@ -0,0 +1,31 @@
+class Cars(carNum: Int) {
+    private val carList = mutableListOf<Car>()
+
+    init {
+        createCar(carNum)
+    }
```
생성자를 통해서 객체 만들기 👍👍
하지만 Car 객체는 이 내부에서 만드는거라 내가 제어할 수 없는 녀석들이 만들어지네!

malibinYun commented on src/main/kotlin/racingcar/Cars.kt:
```diff
@@ -0,0 +1,31 @@
+    private fun createCar(carNum: Int) {
+        for (i in 0 until carNum) {
+            carList.add(Car())
+        }
```
repeat() 함수를 써보는건 어떨까?

malibinYun commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,17 @@
+package racingcar
+
+class Car {
+    private var distanceGraph: String = ""
```
자동차 거리의 상태를 String으로 나타냈다는건, 핵심 도메인 로직과 뷰로직이 합쳐져 있다는 뜻인거 같아.
https://github.com/MalibinStudy/RacingCar/pull/22#discussion_r515570506 이것두 참고해보자

malibinYun commented on src/main/kotlin/racingcar/Race.kt:
```diff
@@ -0,0 +1,25 @@
+package racingcar
+
+class Race(private val carNum: Int, private val raceNum: Int) {
```
raceNum은 무슨 의도인진 알겠지만, 조금 더 구체적인 네이밍이 필요해보여

malibinYun commented on src/main/kotlin/racingcar/Race.kt:
```diff
@@ -0,0 +1,25 @@
+package racingcar
+
+class Race(private val carNum: Int, private val raceNum: Int) {
+    fun start(): String {
+        val cars = Cars(carNum)
+        var fullRecord = ""
+        for (i in 0 until raceNum) {
+            runCars(cars)
+            fullRecord = fullRecord + cars.getDistanceGraph() + NEW_LINE
+        }
+        return fullRecord
+    }
```
오 메서드 명 매우 좋아!
리턴이 String인게 너무 아쉽다. start 했을 때 String이 반환된다는걸 다른 사람이 봤을 땐 String이 어떤 값인지 전혀 모를거같아.

malibinYun replied:
start를 할 때 Cars를 생성하는구나. 아까 처럼 Cars도 내가 컨트롤 하지 못하는 객체가 되어버린거같아. 이러면 Race의 테스트를 진행할 수 있을까?

malibinYun commented on src/main/kotlin/racingcar/RaceRule.kt:
```diff
@@ -0,0 +1,11 @@
+package racingcar
+
+object RaceRule {
+    private val randomRange = IntRange(0, 9)
+    private const val ruleNumber = 4
+
+    fun checkIfCarRun(): Boolean {
+        val randomNum = randomRange.random()
+        return randomNum >= ruleNumber
+    }
+}
```
랜덤 한 값만 뱉어주는 하나의 역할만 하네!
역할 분리 👍👍

malibinYun replied:
상수는 대문자로 하는게 컨벤션이야!

chaeppy replied:
호옹이.. 전 이걸 cars에 다 넣엇는데...반성하게되네요.

malibinYun commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,17 @@
+import java.lang.IllegalArgumentException
+
+object InputView {
+    fun inputCarNum(): Int {
+        println("자동차 대수는 몇 대인가요?")
+        return readLine()?.toInt()
+            ?: throw IllegalArgumentException("자동차 대수는 반드시 입력해야 합니다.")
+    }
```
input보단 get이 더 나아보여!

malibinYun commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,17 @@
+            ?: throw IllegalArgumentException("자동차 대수는 반드시 입력해야 합니다.")
+    }
+
+    fun inputRaceNum(): Int {
+        println("시도할 횟수는 몇 회인가요?")
+        return readLine()?.toInt()
+            ?: throw IllegalArgumentException("시도할 횟수는 반드시 입력해야 합니다.")
+    }
+}
```
readLine()이 null로 떨어지는 경우는 내가 알기로 입력값이 없을 때가 아닌걸로 알고있어.
그래서 입력을 엔터만 쳐도 위의 Exception이 뜨는게 아닌, 공백을 숫자로 바꿀 수 없다는 Exception이 나올거야.

malibinYun reviewed (changes requested):
우선은 테스트가 거의 불가능한 구조인거같아. 그래서 테스트쪽보다는 우선은 도메인과 뷰를 먼저 분리해보자.
테스트가 불가능하다면 좋은 구조가 아니라는 말이있어. 그렇다고 테스트를 위한 코드를 작성하라는 뜻은 아냐! 
분리쪽 다 되면, 특정 값을 넣어주면 특정 값이 나오는걸 테스트하는 것을 잘 생각해보구 내가 컨트롤 할 수 없는 부분들을 컨트롤 할 수 있게 바꿀 수 있게 바꿔보자.

malibinYun commented:
1. 뷰 작성한건 코멘트 몇개 달았어!
2. Car랑 Cars 둘다 직관적이고 좋아! 둘이 엄연히 다른 객체니까 따로 만드는게 맞다고 생각해.
3. input에 대해서는 코멘트를 남겨놨어.
4. 나는 분리한게 훨씬 좋아보여! 자동차가 랜덤한 수를 꺼내는 일 까지 해야되나 싶기도해.

chaeppy commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,17 @@
+    private var distanceGraph: String = ""
+
+    fun run() {
+        if (RaceRule.checkIfCarRun()) {
```
Race 클래스의 runCars에서도 checkIfCarRun을 하는데 car에서도 한번 더 하는 이유가 있을까요..!?!? 궁금합니돠..

RoadtoS7 replied:
실수로 두번 들어 갔어요 반영했습니다~

chaeppy commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,17 @@
+    fun getDistanceGraph() = distanceGraph
+
+    companion object {
+        const val ONE_DISTANCE = "-"
```
아 저도 이거 했어야됐는데 놓쳣네요 역시 빛수현입니다

chaeppy commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -0,0 +1,8 @@
+package racingcar.view
+
+object OutputView {
```
혹씨... InputView 파일은 함수로만 이루어져있는데 OutputView는 오브젝트인 이유가있을까요? just 궁금합니다!!

RoadtoS7 replied:
ObjectView하고 InputView둘다 util성 함수만 존재해서 object타입으로 만들었습니다~

RoadtoS7 committed:
refactor(Car.kt): Fix distance so that it gets separated

도메인과 뷰를 분리했다

RoadtoS7 committed:
refactor(Cars.kt): Fix distance so that it gets separated with view

도메인과 뷰를 분리했다

RoadtoS7 committed:
refactor(InputView.kt): Fix functions name

RoadtoS7 committed:
refactor(InputView.kt): use NumberFormatException instead of IllegalArgumentException

RoadtoS7 committed:
refactor(Race.kt): fix race class to use RaceGraph class and receive cars object so that it doesn't create car instances

RaceGraph로 하여금 뷰를 그리도록 만들었다
그리고 start 메서드에서 Car 인스턴스들을 만드는 것이 아닌 외부에서 생성받아서 생성자를 통해서
받도록 만들었다

RoadtoS7 committed:
refactor(Race.kt): make Race use RaceGraph to draw race result and recieve Car instance so that it doesn't create car instances

RaceGraph에서 뷰를 그리고 기록해 놓도록 만들었다
그리고 start 메서드에서 Car 인스턴스들을 만드는 것이 아닌 외부에서 생성받아서 생성자를 통해서
받도록 만들었다

RoadtoS7 committed:
feat(RaceGraph.kt): add raceGraph for draw race graph

race 결과를 그래프로 그리는 RaceGraph 클래스를 만들었다

RoadtoS7 committed:
refactor(RaceRule.kt): fix constant name so that it satisfies variable naming convention

RoadtoS7 committed:
refactor(RacingCarApplication.kt): fix Race so that it receives car instances and make printRaceResult use raceGraph

RoadtoS7 committed:
refactor(Cars.kt): make run method call each car's run method

run() 메서드가 carIndex를 받아서 특정한 차의 run 메서드만 호출하는 것이 아닌,
모든 차의 run 메서드를 호출하도록 만들었다.

RoadtoS7 committed:
refactor(Race.kt): Remove unnecessary runCars()

SONPYEONGHWA reviewed (commented):
눈이 부신 과제입니다 수현님 고생하셨습니다:)

malibinYun commented on src/main/kotlin/racingcar/Race.kt:
```diff
@@ -0,0 +1,14 @@
+
+import racingcar.view.RaceGraph
+
+class Race(private val cars: Cars, private val raceRound: Int) {
+    fun start(): RaceGraph {
+        val raceGraph = RaceGraph()
+        repeat(raceRound) {
+            cars.run()
+            raceGraph.record(cars)
+        }
+        return raceGraph
+    }
+}
```
Race 객체에서 cars, raceRound를 멤버변수로 받고 생성하는 것 보다
start 메서드에서 그런 정보를 받고 결과를 만들어주기만 하는 것도 좋을듯!

RoadtoS7 replied:
넵 그렇게 해보겠습니다! 
제가 생각한게 맞는지 확인차 여쭤보는 건데욥 start메서드에서 cars와 raceRound를 받았으면 Race 인스턴스가 제가 제어할 수 있는 인스턴스가 된다는 장점이 있는것이 맞나요??

malibinYun replied:
그렇게 생각해볼 수도 있을거같구.
내 의도는 Race객체를 재활용하능하게끔 만드는것도 있구!

malibinYun commented on src/main/kotlin/racingcar/RaceRule.kt:
```diff
@@ -0,0 +1,11 @@
+package racingcar
+
+object RaceRule {
+    private val randomRange = IntRange(0, 9)
```
상수는 대문자 컨벤션!

malibinYun commented on src/main/kotlin/racingcar/view/RaceGraph.kt:
```diff
@@ -0,0 +1,30 @@
+class RaceGraph {
+    private var raceGraph = ""
+
+    fun record(cars: Cars) {
+        val carList = cars.getCars()
+        for (car in carList) {
+            var carDistance = ""
+            repeat(car.getDistance()) {
+                carDistance += ONE_DISTANCE_GRAPH
+            }
+            raceGraph = raceGraph + carDistance + NEW_LINE
+        }
+        raceGraph += NEW_LINE
+    }
```
여기는 뷰를 실제로 그리는 부분인거같아. 이것도 도메인과 뷰가 섞인 느낌이야.
뷰를 그리는건 뷰에서만 해보자!

RoadtoS7 replied:
도메인과 뷰를 분리하기 위해서 일부러 뷰를 그리는 RaceGraph라는 아이를 만들었습니다. 
그래서 RaceGraph를 Race 클래스의 start() 메서드에서 호출했는데, 이렇게 해도 도메인과 뷰가 섞인 방법인지 궁금합니다ㅠㅠ

malibinYun replied:
inputView outputView 쪽을 다 떼내고 로직들을 안드로이드로 옮긴다고 생각해보자
얘는 안드로이드에 보여줄 View들을 만드는게 아니라, 그저 String 들이 합쳐져있는 문자열 덩어리일뿐이잖아? 그러면 이 뷰는 안드로이드에 포팅할 때 완벽하게 싹 뜯어고쳐야하는 뷰가 되겠지?

start()를 하면 Record를 한 객체를 뱉게 한 후에, View에서 그 객체를 읽어서 출력한다면 어떨까?

malibinYun commented on src/main/kotlin/racingcar/view/RaceGraph.kt:
```diff
@@ -0,0 +1,30 @@
+        val carList = cars.getCars()
+        for (car in carList) {
+            var carDistance = ""
+            repeat(car.getDistance()) {
+                carDistance += ONE_DISTANCE_GRAPH
+            }
```
요기 인덴트 초과!

malibinYun reviewed (approved):
자 이제 오늘 배운거 반영이랑 다음단계를 진행해봅시다!
